sap-service-operator/src/main/java/com/sap/cloud/environment/servicebinding/FileSystemWatcherCache.java

```diff
@@ -0,0 +1,155 @@
+        return directories
+            .stream()
+            .peek(this::renewCachedServiceBindingIfNeeded)
+            .map(cachedServiceBindings::get)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+    }
+
+    private void removeOutdatedWatchKeys( @Nonnull final Collection<Path> directoriesOfInterest )
+    {
+        for( final Map.Entry<Path, WatchKey> entry : directoryWatchKeys.entrySet() ) {
+            if( directoriesOfInterest.contains(entry.getKey()) ) {
+                continue;
+            }
+
+            entry.getValue().cancel();
+            directoryWatchKeys.remove(entry.getKey());
+        }
+    }
+
+    private void removeOutdatedServiceBindings( @Nonnull final Collection<Path> directoriesOfInterest )
+    {
+        for( final Map.Entry<Path, ServiceBinding> entry : cachedServiceBindings.entrySet() ) {
+            if( directoriesOfInterest.contains(entry.getKey()) ) {
+                continue;
+            }
+
+            cachedServiceBindings.remove(entry.getKey());
+        }
+    }
+
+    private void renewCachedServiceBindingIfNeeded( @Nonnull final Path directory )
+    {
+        @Nullable
+        final WatchKey watchKey = directoryWatchKeys.get(directory);
+        if( watchKey == null ) {
+            watchAndCacheServiceBinding(directory);
+            return;
+        }
+
+        if( !watchKey.isValid() ) {
+            throw new IllegalStateException(
+                String.format("%s for directory '%s' is invalid.", WatchKey.class.getSimpleName(), directory));
+        }
+
+        if( hasBeenModified(watchKey) ) {
+            cacheServiceBinding(directory);
+        }
+    }
+
+    private void watchAndCacheServiceBinding( @Nonnull final Path directory )
+    {
+        startWatching(directory);
+        cacheServiceBinding(directory);
+    }
+
+    private void cacheServiceBinding( @Nonnull final Path directory )
+    {
+        cachedServiceBindings.remove(directory);
+
+        @Nullable
+        final ServiceBinding serviceBinding = serviceBindingLoader.apply(directory);
+        if( serviceBinding == null ) {
+            return;
+        }
+
+        cachedServiceBindings.put(directory, serviceBinding);
+    }
+
+    private void startWatching( @Nonnull final Path directory )
+    {
+        try {
+            final WatchKey watchKey = directory.register(watchService, ENTRY_CREATE, ENTRY_MODIFY, ENTRY_DELETE);
+            directoryWatchKeys.put(directory, watchKey);
+        }
+        catch( final IOException e ) {
+            throw new IllegalStateException(String.format("Unable to watch directory '%s'.", directory), e);
+        }
+    }
+
+    private boolean hasBeenModified( @Nonnull final WatchKey watchKey )
+    {
+        final List<WatchEvent<?>> events = watchKey.pollEvents();
+        watchKey.reset();
+
+        return events.stream().map(WatchEvent::kind).anyMatch(MODIFICATION_EVENTS::contains);
```

Comment:
(Question)
I'm confused: with the `register` method above, you already added a filter or not?
```suggestion
        return !events.isEmpty();
```

Reply:
That's true. However, the `WatchKey` might always capture events of type `OVERFLOW` (even if not registered for that event type).